liquidctl/driver/msi.py

```diff
@@ -0,0 +1,971 @@
+    COLOR_PULSE = 21
+    COLOR_RING = 15
+    COLOR_RING_DOUBLE_FLASHING = 35
+    COLOR_RING_FLASHING = 34
+    COLOR_SHIFT = 22
+    COLOR_WAVE = 23
+    CORSAIR_IQUE = 37
+    DISABLE = 0
+    DISABLE2 = 33
+    DOUBLE_FLASHING = 4
+    DOUBLE_METEOR = 17
+    ENERGY = 18
+    FAN_CONTROL = 32
+    FIRE = 38
+    FLASHING = 3
+    JAZZ = 12
+    JRAINBOW = 28
+    LAVA = 39
+    LIGHTING = 5
+    MARQUEE = 24
+    METEOR = 7
+    MOVIE = 14
+    MSI_MARQUEE = 6
+    MSI_RAINBOW = 9
+    NO_ANIMATION = 1
+    PLANETARY = 16
+    PLAY = 13
+    POP = 10
+    RAINBOW = 25
+    RAINBOW_DOUBLE_FLASHING = 30
+    RAINBOW_FLAHING = 29
+    RAINBOW_WAVE = 26
+    RANDOM = 31
+    RAP = 11
+    STACK = 36
+    VISOR = 27
+    WATER_DROP = 8
+    END = 40
+
+
+@unique
+class _Speed(RelaxedNamesEnum):
+    LOW = 0
+    MEDIUM = 1
+    HIGH = 2
+
+
+@unique
+class _ColorSelection(Enum):
+    RAINBOW_COLOR = 0
+    USER_DEFINED = 1
+
+
+@unique
+class _JType(Enum):
+    JRAINBOW = 3
+    JCORSAIR = 4
+    JONBOARD = 5
+
+
+@unique
+class _UploadType(Enum):
+    BANNER = 0
+    GIF = 1
+
+
+class MpgCooler(UsbHidDriver):
+    _MATCHES = [
+        (0x0db0, 0xb130, 'MSI MPG Coreliquid K360', {
+            'fan_count': 5
+        }),
+        (0x0db0, 0xca00, 'Unknown', {}),
+        (0x0db0, 0xca02, 'Unknown', {}),
+    ]
+
+    def __init__(self, device, description, **kwargs):
+        super().__init__(device, description, **kwargs)
+        self._feature_data_per_led = bytearray(_PER_LED_LENGTH + 5)
+        self._bytearray_oled_hardware_monitor_data = bytearray(_REPORT_LENGTH)
+        self._per_led_rgb_jonboard = bytearray(_PER_LED_LENGTH)
+        self._per_led_rgb_jrainbow1 = bytearray(_PER_LED_LENGTH)
+        self._per_led_rgb_jrainbow2 = bytearray(_PER_LED_LENGTH)
+        self._per_led_rgb_jcorsair = bytearray(_PER_LED_LENGTH)
+        self._fan_count = kwargs.pop('fan_count', 0)
+        # the following fields are only initialized in connect()
+        self._data = None
+        self._feature_data = None
+
+    @classmethod
+    def probe(cls, handle, vendor=None, product=None, release=None,
+              serial=None, match=None, **kwargs):
+        for vid, pid, desc, devargs in cls._MATCHES:
+            if (vendor and vendor != vid) or handle.vendor_id != vid:
+                continue
+            if (product and product != pid) or handle.product_id != pid:
+                continue
+            if release and handle.release_number != release:
+                continue
+            if serial and handle.serial_number != serial:
+                continue
+            if match and match.lower() not in desc.lower():
+                continuei
+            if handle.hidinfo['usage_page'] != 0xff00:
+                continue
+            consargs = devargs.copy()
+            consargs.update(kwargs)
+            dev = cls(handle, desc, **consargs)
+            _LOGGER.debug('%s identified: %s', cls.__name__, desc)
+            yield dev
```

Comment:
Take a look at

https://github.com/liquidctl/liquidctl/blob/3c227276666dbfee2b60491dbe3d87b35598f776/liquidctl/driver/rgb_fusion2.py#L107-L126

for an example of how this might be simplified and also made a bit more robust.